### PR TITLE
[#IC-388]  Add new fn-app-messages client + A/B testing

### PR DESCRIPTION
**Pull request by gquadrati**
This PR depends on https://github.com/pagopa/io-infra/pull/142

#### List of Changes

- add new client for fn-app-messages
- add messagesServiceSelector for choosing between services (fn-app and fn-app-messages)

#### Motivation and Context

Switch to new fn-app-messages function

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -404,6 +411,10 @@ export function newApp({
         API_CLIENT,
         new PecServerClientFactory(PECSERVERS)
       );
+      // Create the new messages service.
+      const NEW_MESSAGES_SERVICE = new NewMessagesService(
```
Comment (AleDore) [inline src/app.ts:415]: APP_MESSAGES_SERVICE?

File: src/config.ts [hunk 2]
```diff
@@ -360,11 +361,44 @@ export const getHttpsApiFetchWithBearer = (bearer: string) =>
     )
   );
 
+export const MessagesFeatureFlagType = t.union([
```
Comment (gunzip) [inline src/config.ts:364]: https://github.com/gcanti/io-ts/tree/1.x#union-of-string-literals

File: src/controllers/messagesController.ts [hunk 3]
```diff
@@ -101,11 +107,13 @@ export default class MessagesController {
     withUserFromRequest(req, user =>
       tryCatch(
         () =>
-          this.messagesService.getLegalMessage(
-            user,
-            req.params.id,
-            this.tokenService.getPecServerTokenHandler(user.fiscal_code)
-          ),
+          this.messageServicSelector
+            .getOldMessageService()
```
Comment (gunzip) [inline src/controllers/messagesController.ts:111]: may you add a comment on why we don't use `select()` here?

File: src/controllers/messagesController.ts [hunk 4]
```diff
@@ -121,12 +129,14 @@ export default class MessagesController {
...
       tryCatch(
         () =>
-          this.messagesService.getLegalMessageAttachment(
-            user,
-            req.params.id,
-            this.tokenService.getPecServerTokenHandler(user.fiscal_code),
-            req.params.attachment_id
-          ),
+          this.messageServicSelector
+            .getOldMessageService()
```
Comment (gunzip) [inline src/controllers/messagesController.ts:133]: ditto

File: src/controllers/messagesController.ts [hunk 5]
```diff
@@ -72,7 +73,10 @@ export default class MessagesController {
           minimumId: req.query.minimum_id
           /* eslint-enable sort-keys */
         }),
-        params => this.messagesService.getMessagesByUser(user, params)
+        params =>
+          this.messageServicSelector
```
Comment (AleDore) [inline src/controllers/messagesController.ts:77]: typo

File: src/config.ts [hunk 6]
```diff
@@ -360,11 +361,44 @@ export const getHttpsApiFetchWithBearer = (bearer: string) =>
...
+
+export const FF_MESSAGES_TYPE = MessagesFeatureFlagType.decode(
+  process.env.FF_MESSAGES_TYPE
+).getOrElse("none");
+
+export const FF_MESSAGES_BETA_TESTER_LIST = CommaSeparatedListOf(NonEmptyString)
+  .decode(process.env.FF_MESSAGES_BETA_TESTER_LIST ?? "")
+  .getOrElse([]);
+
+export const FF_MESSAGES_CARARY_USERS_REGEX = NonEmptyString.decode(
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply src/config.ts]: Done in [d951fe2](https://github.com/pagopa/io-backend/pull/882/commits/d951fe237e764033c912db07a0b8f8252f376576)

File: src/app.ts (hunk 1 above)
Comment (gquadrati) [reply src/app.ts]: Done in [d951fe2](https://github.com/pagopa/io-backend/pull/882/commits/d951fe237e764033c912db07a0b8f8252f376576)

File: src/config.ts (hunk 2 above)
Comment (gquadrati) [reply src/config.ts]: Thanks! Done in [f0c9eca](https://github.com/pagopa/io-backend/pull/882/commits/f0c9ecac61390c067788b47867ae612207dbd68c)

File: src/controllers/messagesController.ts (hunk 3 above)
Comment (gquadrati) [reply src/controllers/messagesController.ts:113]: Done in [f0c9eca](https://github.com/pagopa/io-backend/pull/882/commits/f0c9ecac61390c067788b47867ae612207dbd68c)

File: src/controllers/messagesController.ts (hunk 5 above)
Comment (gquadrati) [reply src/controllers/messagesController.ts]: Done in [f0c9eca](https://github.com/pagopa/io-backend/pull/882/commits/f0c9ecac61390c067788b47867ae612207dbd68c)

**Review by AleDore — APPROVED** [review]
LGTM

**Review by AleDore — APPROVED** [review]
LGTM

**Pull request merged**